File: services/api-gateway/src/utils/circuitBreaker.js
class CircuitBreaker {
  constructor(name, options = {}) {
    this.name = name;
    this.failureThreshold = options.failureThreshold || 5;
    this.recoveryTimeoutMs = options.recoveryTimeoutMs || 30000;
    this.successThreshold = options.successThreshold || 2;

    this.state = 'CLOSED';
    this.failures = 0;
    this.halfOpenSuccesses = 0;
    this.openedAt = null;
  }

  canRequest() {
    if (this.state === 'CLOSED') {
      return true;
    }

    if (this.state === 'OPEN') {
      const now = Date.now();
      if (this.openedAt && now - this.openedAt >= this.recoveryTimeoutMs) {
        this.state = 'HALF_OPEN';
        this.halfOpenSuccesses = 0;
        return true;
      }
      return false;
    }

    return true;
  }

  onSuccess() {
    if (this.state === 'HALF_OPEN') {
      this.halfOpenSuccesses += 1;
      if (this.halfOpenSuccesses >= this.successThreshold) {
        this.state = 'CLOSED';
        this.failures = 0;
        this.halfOpenSuccesses = 0;
      }
      return;
    }

    this.failures = 0;
  }

  onFailure() {
    if (this.state === 'HALF_OPEN') {
      this.tripOpen();
      return;
    }

    this.failures += 1;
    if (this.failures >= this.failureThreshold) {
      this.tripOpen();
    }
  }

  tripOpen() {
    this.state = 'OPEN';
    this.openedAt = Date.now();
    this.failures = 0;
    this.halfOpenSuccesses = 0;
  }
}

module.exports = CircuitBreaker;

File: services/ai-service/app/utils/circuit_breaker.py
import time
from typing import Any, Awaitable, Callable, Optional


class AsyncCircuitBreaker:
    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout_seconds: int = 30,
        half_open_success_threshold: int = 2,
    ) -> None:
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout_seconds = recovery_timeout_seconds
        self.half_open_success_threshold = half_open_success_threshold

        self.state = "CLOSED"
        self.failures = 0
        self.half_open_successes = 0
        self.opened_at: Optional[float] = None

    def can_execute(self) -> bool:
        if self.state == "CLOSED":
            return True

        if self.state == "OPEN":
            now = time.time()
            if self.opened_at and (now - self.opened_at) >= self.recovery_timeout_seconds:
                self.state = "HALF_OPEN"
                self.half_open_successes = 0
                return True
            return False

        return True

    def on_success(self) -> None:
        if self.state == "HALF_OPEN":
            self.half_open_successes += 1
            if self.half_open_successes >= self.half_open_success_threshold:
                self.state = "CLOSED"
                self.failures = 0
                self.half_open_successes = 0
            return

        self.failures = 0

    def on_failure(self) -> None:
        if self.state == "HALF_OPEN":
            self.trip_open()
            return

        self.failures += 1
        if self.failures >= self.failure_threshold:
            self.trip_open()

    def trip_open(self) -> None:
        self.state = "OPEN"
        self.opened_at = time.time()
        self.failures = 0
        self.half_open_successes = 0

    async def call(
        self,
        fn: Callable[[], Awaitable[Any]],
        fallback: Optional[Callable[[], Any]] = None,
    ) -> Any:
        if not self.can_execute():
            if fallback:
                return fallback()
            raise RuntimeError(f"circuit open for {self.name}")

        try:
            result = await fn()
            self.on_success()
            return result
        except Exception:
            self.on_failure()
            if fallback:
                return fallback()
            raise
